pdns/dnswriter.cc

```diff
@@ -183,145 +162,198 @@ void DNSPacketWriter::xfr8BitInt(uint8_t val)
 {
   if(text.empty()) {
-    d_record.push_back(0);
+    d_content.push_back(0);
     return;
   }
   if(lenField)
-    d_record.push_back(text.length());
-  d_record.insert(d_record.end(), text.c_str(), text.c_str() + text.length());
+    d_content.push_back(text.length());
+  d_content.insert(d_content.end(), text.c_str(), text.c_str() + text.length());
 }
 
-/* FIXME400: check that this beats a map */
-DNSPacketWriter::lmap_t::iterator find(DNSPacketWriter::lmap_t& nmap, const DNSName& name)
+
+static constexpr bool l_verbose=false;
+uint16_t DNSPacketWriter::lookupName(const DNSName& name, uint16_t* matchLen)
 {
-  DNSPacketWriter::lmap_t::iterator ret;
-  for(ret=nmap.begin(); ret != nmap.end(); ++ret)
-    if(ret->first == name)
-      break;
-  return ret;
-}
+  // iterate over the written labels, see if we find a match
+  const auto& raw = name.getStorage();
 
-// //! tokenize a label into parts, the parts describe a begin offset and an end offset
-// bool labeltokUnescape(labelparts_t& parts, const string& label)
-// {
-//   string::size_type epos = label.size(), lpos(0), pos;
-//   bool unescapedSomething = false;
-//   const char* ptr=label.c_str();
-
-//   parts.clear();
-
-//   for(pos = 0 ; pos < epos; ++pos) {
-//     if(ptr[pos]=='\\') {
-//       pos++;
-//       unescapedSomething = true;
-//       continue;
-//     }
-//     if(ptr[pos]=='.') {
-//       parts.push_back(make_pair(lpos, pos));
-//       lpos=pos+1;
-//     }
-//   }
-
-//   if(lpos < pos)
-//     parts.push_back(make_pair(lpos, pos));
-//   return unescapedSomething;
-// }
+  /* name might be a.root-servers.net, we need to be able to benefit from finding:
+     b.root-servers.net, or even:
+     b\xc0\x0c 
+  */
+  unsigned int bestpos=0;
+  *matchLen=0;
+  boost::container::static_vector<uint16_t, 34> nvect, pvect;
+
+  try {
+    for(auto riter= raw.cbegin(); riter < raw.cend(); ) {
+      if(!*riter)
+        break;
+      nvect.push_back(riter - raw.cbegin());
+      riter+=*riter+1;
+    }
+  }
+  catch(std::bad_alloc& ba) {
+    if(l_verbose)
+      cout<<"Domain "<<name<<" too large to compress"<<endl;
+    return 0;
+  }
+  
+  if(l_verbose) {
+    cout<<"Input vector for lookup "<<name<<": ";
+    for(const auto n : nvect) 
+      cout << n<<" ";
+    cout<<endl;
+    cout<<makeHexDump(string(raw.c_str(), raw.c_str()+raw.size()))<<endl;
+  }
 
+  if(l_verbose)
+    cout<<"Have "<<d_namepositions.size()<<" to ponder"<<endl;
+  int counter=1;
+  for(auto p : d_namepositions) {
+    if(l_verbose) {
+      cout<<"Pos: "<<p<<", "<<d_content.size()<<endl;
+      DNSName pname((const char*)&d_content[0], d_content.size(), p, true); // only for debugging
+      cout<<"Looking at '"<<pname<<"' in packet at position "<<p<<"/"<<d_content.size()<<", option "<<counter<<"/"<<d_namepositions.size()<<endl;
+      ++counter;
+    }
+    // memcmp here makes things _slower_
+    pvect.clear();
+    try {
```

Comment:
I wonder if we could compute `pvect` once when inserting a new entry into `d_namepositions` and store it instead.